src/PostgREST/Auth.hs

```diff
@@ -177,24 +179,49 @@ middleware appState app req respond = do
 -- | Used to retrieve and insert JWT to JWT Cache
 getJWTFromCache :: AppState -> ByteString -> Int -> IO (Either Error AuthResult) -> UTCTime -> IO (Either Error AuthResult)
 getJWTFromCache appState token maxLifetime parseJwt utc = do
-  checkCache <- C.lookup (getJwtCache appState) token
+
+  checkCache <- C.lookup jwtCache token
   authResult <- maybe parseJwt (pure . Right) checkCache
 
+  -- if token not found, add to cache and increment cache size metric
   case (authResult,checkCache) of
-    (Right res, Nothing) -> C.insert' (getJwtCache appState) (getTimeSpec res maxLifetime utc) token res
+    (Right res, Nothing) -> do
+      let tSpec = getTimeSpec res maxLifetime utc
+      C.insert' jwtCache (Just tSpec) token res
+      entrySize <- calcCacheEntrySizeInBytes (token, res, tSpec) -- adds to cache
+      observer $ JWTCache entrySize
+
     _                    -> pure ()
 
   return authResult
+    where
+      observer = getObserver appState
+      jwtCache = getJwtCache appState
 
 -- Used to extract JWT exp claim and add to JWT Cache
-getTimeSpec :: AuthResult -> Int -> UTCTime -> Maybe TimeSpec
+getTimeSpec :: AuthResult -> Int -> UTCTime -> TimeSpec
 getTimeSpec res maxLifetime utc = do
   let expireJSON = KM.lookup "exp" (authClaims res)
       utcToSecs = floor . nominalDiffTimeToSeconds . utcTimeToPOSIXSeconds
       sciToInt = fromMaybe 0 . Sci.toBoundedInteger
   case expireJSON of
-    Just (JSON.Number seconds) -> Just $ TimeSpec (sciToInt seconds - utcToSecs utc) 0
-    _                          -> Just $ TimeSpec (fromIntegral maxLifetime :: Int64) 0
+    Just (JSON.Number seconds) -> TimeSpec (sciToInt seconds - utcToSecs utc) 0
+    _                          -> TimeSpec (fromIntegral maxLifetime :: Int64) 0
+
+-- | Calculate a single entry of JWT Cache Size in Bytes
+calcCacheEntrySizeInBytes :: (ByteString, AuthResult, TimeSpec) -> IO Int
```

Comment:
Hm, how about a type alias for `(ByteString, AuthResult, TimeSpec)`?

Reply:
Hm, how about
```haskell
type CacheEntry = (ByteString, AuthResult, TimeSpec)
```
?

Reply:
Maybe even a `data` type to know which part each type represents.

Reply:
Oh, we can do this:
```haskell
data CacheEntry
  = CacheKey ByteString
  | CacheValue AuthResult
  | ExpireVal TimeSpec
```

Reply:
@steve-chavez Actually, could we skip the type alias in this PR? Actually, I am avoiding introducing this because of following reasons:

- We are polluting `Auth.hs`
- I am actually planning to introduce `Cache.hs` module in #3801. This module will include all cache related logic and types

Reply:
Ah, it's getting clearer. But I thought all of those are part of the entry? Wouldn't it be like:

```hs
data CacheEntry = { key :... , value:..., expireVal: ...}
```

Reply:
Oh, sorry, yes it should be a record.